sys/include/net/sock/dtls.h

```diff
@@ -0,0 +1,618 @@
+ *
+ * @{
+ *
+ * @file
+ * @brief   DTLS sock definitions
+ *
+ * @author  Aiman Ismail <muhammadaimanbin.ismail@haw-hamburg.de>
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ * @author  Raul A. Fuentes Samaniego <raul.fuentes-samaniego@inria.fr>
+ * @author  Daniele Lacamera <daniele@wolfssl.com>
+ * @author  Ken Bannister <kb2ma@runbox.com>
+ */
+
+#ifndef NET_SOCK_DTLS_H
+#define NET_SOCK_DTLS_H
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/types.h>
+
+#include "net/sock.h"
+#include "net/sock/udp.h"
+#include "net/credman.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name DTLS version number
+ * @anchor sock_dtls_prot_version
+ * @{
+ */
+#define DTLSv1_0    (1)    /**< DTLS version 1.0 */
+#define DTLSv1_2    (2)    /**< DTLS version 1.2 */
+/** @} */
+
+/**
+ * @name DTLS role
+ * @anchor sock_dtls_role
+ * @{
+ */
+#define DTLS_CLIENT (1),    /**< Endpoint client role */
+#define DTLS_SERVER (2),    /**< Endpoint server role */
+/** @} */
+
+/**
+ * @brief Method of connecting to remote
+ */
+typedef struct {
+    /**
+     * @brief DTLS version number
+     *
+     * @see [DTLS version number](@ref sock_dtls_prot_version)
+     */
+    uint8_t dtls_version;
+
+    /**
+     * @brief DTLS role
+     *
+     * @see [DTLS role](@ref sock_dtls_role)
+     */
+    uint8_t role;
+} sock_dtls_method_t;
+
+/**
+ * @brief   Type for a DTLS sock object
+ *
+ * @note    API implementors: `struct sock_dtls` needs to be defined by
+ *          an implementation-specific `sock_dtls_types.h`.
+ */
+typedef struct sock_dtls sock_dtls_t;
+
+/**
+ * @brief Information about an established session with a remote endpoint.
+ */
+typedef struct sock_dtls_session sock_dtls_session_t;
+
+/**
+ * @brief Called exactly once during autoinit.
+ *
+ * Calls the initialization function required by the DTLS stack used.
+ */
+void sock_dtls_init(void);
+
+/**
+ * @brief Creates a new DTLS sock object
+ *
+ * Takes an initialized UDP sock and uses it for the transport.
+ * Memory allocation functions required by the underlying DTLS
+ * stack can be called in this function.
+ *
+ * @see net_credman.
+ *
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock initialized with
+ *                      @ref sock_udp_create()to be used underneath.
+ * @param[in] tag       Credential tag of the sock. The sock will only use
+ *                      credentials with the same tag given here.
+ * @param[in] method    Defines the role of the endpoint and the DTLS version
+ *                      to use.
+ *
+ * @return  0 on success.
+ * @return  -1 on error
+ */
+int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
+                     credman_tag_t tag, sock_dtls_method_t method);
+
+/**
+ * @brief Initialises the server to listen for incoming connections
+ *
+ * @param[in] sock      DTLS sock to listen to
+ */
+void sock_dtls_init_server(sock_dtls_t *sock);
```

Comment:
Oh, and this is not needed anymore, right?

Reply:
Yup, I'll remove it in the next commit.

Reply:
Done.